File: platform/android/lynx_android/src/main/java/com/lynx/tasm/gesture/handler/PanGestureHandler.java
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.gesture.handler;

import android.view.MotionEvent;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.lynx.react.bridge.ReadableMap;
import com.lynx.tasm.behavior.LynxContext;
import com.lynx.tasm.event.LynxTouchEvent;
import com.lynx.tasm.gesture.GestureArenaMember;
import com.lynx.tasm.gesture.common.GestureExtraBundle;
import com.lynx.tasm.gesture.detector.GestureDetector;
import com.lynx.tasm.utils.PixelUtils;
import java.util.HashMap;

/**
 * The PanGestureHandler class is a concrete implementation of the BaseGestureHandler class
 * for handling pan gestures. It handles pan gesture for component, triggered when a finger touches
 * the component.
 */
public class PanGestureHandler extends BaseGestureHandler {
  private final HashMap<String, Object> mEventParams;

  // Min distance in X and Y axis,  If the finger travels further than the defined distance
  // along the X axis and the gesture will activated.
  // otherwise assigned to 0
  private float mMinDistance = PixelUtils.dipToPx(0);

  // record x axis when action down
  private float mStartX = 0f;
  // record y axis when action down
  private float mStartY = 0f;
  // record x axis when action move / up
  private float mLastX = 0f;
  // record x axis when action move / up
  private float mLastY = 0f;

  // record last lynx touch event
  private LynxTouchEvent mLastTouchEvent;

  // is onBegin invoked or not
  private boolean mIsInvokedBegin = false;

  // is onStart invoked or not
  private boolean mIsInvokedStart = false;

  // is onEnd invoked or not
  private boolean mIsInvokedEnd = false;

  /**
   * Constructs a PanGestureHandler object with the specified properties.
   *
   * @param sign                the sign indicating the direction of the gesture
   * @param lynxContext         the LynxContext associated with the gesture handler
   * @param gestureDetector     the GestureDetector associated with the gesture handler
   * @param gestureArenaMember  the GestureArenaMember associated with the gesture handler
   */
  public PanGestureHandler(int sign, LynxContext lynxContext,
      @NonNull GestureDetector gestureDetector, GestureArenaMember gestureArenaMember) {
    super(sign, lynxContext, gestureDetector, gestureArenaMember);
    handleConfigMap(gestureDetector.getConfigMap());
    mEventParams = new HashMap<>();
  }

  @Override
  protected void handleConfigMap(@Nullable ReadableMap config) {
    if (config == null) {
      return;
    }
    mMinDistance = PixelUtils.dipToPx(config.getLong(GestureConstants.MIN_DISTANCE, 0), 0);
  }

  @Override
  protected void onHandle(@Nullable MotionEvent event, @Nullable LynxTouchEvent lynxTouchEvent,
      float flingDeltaX, float flingDeltaY, boolean handleBySimultaneous,
      @Nullable GestureExtraBundle extraBundle) {
    if (lynxTouchEvent != null) {
      mLastTouchEvent = lynxTouchEvent;
    }
    // If the event is empty, it means the finger not touches the screen
    if (event == null) {
      ignore();
      return;
    }
    if (mStatus >= GestureConstants.LYNX_STATE_FAIL) {
      return;
    }
    switch (event.getActionMasked()) {
      case MotionEvent.ACTION_DOWN:
        mStartX = event.getX();
        mStartY = event.getY();
        mIsInvokedEnd = false;
        begin();
        onBegin(mStartX, mStartY, lynxTouchEvent);
        break;
      case MotionEvent.ACTION_MOVE:
        mLastX = event.getX();
        mLastY = event.getY();
        if (mStatus == GestureConstants.LYNX_STATE_INIT) {
          begin();
          onBegin(mLastX, mLastY, lynxTouchEvent);
        }

        if (shouldActive()) {
          activate();
          onStart(mLastX, mStartY, lynxTouchEvent);
        }
        if (mStatus == GestureConstants.LYNX_STATE_ACTIVE) {
          onUpdate(mLastX, mLastY, lynxTouchEvent, extraBundle);
        } else if (mStatus >= GestureConstants.LYNX_STATE_FAIL) {
          onEnd(mLastX, mLastY, lynxTouchEvent);
        }
        break;
      case MotionEvent.ACTION_UP:
        fail();
        onEnd(mLastX, mLastY, mLastTouchEvent);
        break;
      default:
        break;
    }
  }

  private boolean shouldActive() {
    if (mStatus >= GestureConstants.LYNX_STATE_FAIL) {
      return false;
    }
    float dx = Math.abs(mLastX - mStartX);
    float dy = Math.abs(mLastY - mStartY);
    if (dx > mMinDistance || dy > mMinDistance) {
      return true;
    }
    return false;
  }

  @Override
  public void fail() {
    super.fail();
    onEnd(mLastX, mLastY, mLastTouchEvent);
  }

  @Override
  public void end() {
    super.end();
    onEnd(mLastX, mLastY, mLastTouchEvent);
  }

  @Override
  public void reset() {
    super.reset();
    mIsInvokedBegin = false;
    mIsInvokedStart = false;
    mIsInvokedEnd = false;
  }

  protected HashMap<String, Object> getEventParamsInActive(@Nullable LynxTouchEvent event) {
    mEventParams.put("scrollX", px2dip(mGestureArenaMember.getMemberScrollX()));
    mEventParams.put("scrollY", px2dip(mGestureArenaMember.getMemberScrollY()));
    mEventParams.put("isAtStart", mGestureArenaMember.isAtBorder(true));
    mEventParams.put("isAtEnd", mGestureArenaMember.isAtBorder(false));
    mEventParams.putAll(getEventParamsFromTouchEvent(event));
    return mEventParams;
  }

  @Override
  protected void onBegin(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnBeginEnable() || mIsInvokedBegin) {
      return;
    }
    mIsInvokedBegin = true;
    sendGestureEvent(GestureConstants.ON_BEGIN, getEventParamsInActive(event));
  }

  @Override
  protected void onUpdate(float deltaX, float deltaY, @Nullable LynxTouchEvent event,
      @Nullable GestureExtraBundle extraBundle) {
    if (!isOnUpdateEnable()) {
      return;
    }
    sendGestureEvent(GestureConstants.ON_UPDATE, getEventParamsInActive(event));
  }

  @Override
  protected void onStart(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnStartEnable() || mIsInvokedStart || !mIsInvokedBegin) {
      return;
    }
    mIsInvokedStart = true;
    sendGestureEvent(GestureConstants.ON_START, getEventParamsInActive(event));
  }

  @Override
  protected void onEnd(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnEndEnable() || mIsInvokedEnd || !mIsInvokedBegin) {
      return;
    }
    mIsInvokedEnd = true;
    sendGestureEvent(GestureConstants.ON_END, getEventParamsInActive(event));
  }
}

File: platform/harmony/lynx_harmony/src/main/cpp/gesture/handler/pan_gesture_handler.cc
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "platform/harmony/lynx_harmony/src/main/cpp/gesture/handler/pan_gesture_handler.h"

#include "platform/harmony/lynx_harmony/src/main/cpp/event/gesture_event.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/event/touch_event.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/gesture/gesture_arena_member.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/lynx_context.h"

namespace lynx {
namespace tasm {
namespace harmony {
PanGestureHandler::PanGestureHandler(
    int sign, LynxContext* lynx_context,
    std::shared_ptr<GestureDetector> gesture_detector,
    std::weak_ptr<GestureArenaMember> gesture_arena_member)
    : BaseGestureHandler(sign, lynx_context, gesture_detector,
                         gesture_arena_member),
      min_distance_(0),
      start_x_(0),
      start_y_(0),
      last_x_(0),
      last_y_(0),
      is_invoked_begin_(false),
      is_invoked_start_(false),
      is_invoked_end_(false) {
  HandleConfigMap(gesture_detector->gesture_config());
}

void PanGestureHandler::HandleConfigMap(const lepus::Value& config) {
  if (config.Contains(GestureConstants::MIN_DISTANCE)) {
    min_distance_ =
        Dip2Px(config.GetProperty(GestureConstants::MIN_DISTANCE).Number());
  }
}

void PanGestureHandler::OnHandle(
    const ArkUI_UIInputEvent* event,
    const std::shared_ptr<TouchEvent>& lynx_touch_event, float fling_delta_x,
    float fling_delta_y, bool handle_by_simultaneous,
    const std::shared_ptr<GestureExtraBundle>& extra_bundle) {
  last_touch_event_ = lynx_touch_event;
  if (event == nullptr) {
    Ignore();
    return;
  }
  if (status_ >= GestureConstants::LYNX_STATE_FAIL) {
    return;
  }

  switch (OH_ArkUI_UIInputEvent_GetAction(event)) {
    case UI_TOUCH_EVENT_ACTION_DOWN:
      start_x_ = OH_ArkUI_PointerEvent_GetXByIndex(event, 0);
      start_y_ = OH_ArkUI_PointerEvent_GetYByIndex(event, 0);
      is_invoked_end_ = false;
      Begin();
      OnBegin(start_x_, start_y_, lynx_touch_event);
      break;
    case UI_TOUCH_EVENT_ACTION_MOVE:
      last_x_ = OH_ArkUI_PointerEvent_GetXByIndex(event, 0);
      last_y_ = OH_ArkUI_PointerEvent_GetYByIndex(event, 0);
      if (status_ == GestureConstants::LYNX_STATE_INIT) {
        Begin();
        OnBegin(last_x_, last_y_, lynx_touch_event);
      }
      if (ShouldActivate()) {
        Activate();
        OnStart(last_x_, start_y_, lynx_touch_event);
      }
      if (status_ == GestureConstants::LYNX_STATE_ACTIVE) {
        OnUpdate(last_x_, last_y_, lynx_touch_event, extra_bundle);
      } else if (status_ >= GestureConstants::LYNX_STATE_FAIL) {
        OnEnd(last_x_, last_y_, lynx_touch_event);
      }
      break;
    case UI_TOUCH_EVENT_ACTION_UP:
      Fail();
      OnEnd(last_x_, last_y_, lynx_touch_event);
      break;
    default:
      break;
  }
}

bool PanGestureHandler::ShouldActivate() {
  float dx = std::abs(last_x_ - start_x_);
  float dy = std::abs(last_y_ - start_y_);
  return dx > min_distance_ || dy > min_distance_;
}

void PanGestureHandler::Fail() {
  BaseGestureHandler::Fail();
  OnEnd(last_x_, last_y_, last_touch_event_);
}

void PanGestureHandler::End() {
  BaseGestureHandler::End();
  OnEnd(last_x_, last_y_, last_touch_event_);
}

void PanGestureHandler::Reset() {
  BaseGestureHandler::Reset();
  is_invoked_begin_ = false;
  is_invoked_start_ = false;
  is_invoked_end_ = false;
}

lepus::Value PanGestureHandler::GetEventParamsInActive(
    const std::shared_ptr<TouchEvent>& touch_event) {
  auto event_params_ = lepus::Dictionary::Create();

  event_params_->SetValue("scrollX", gesture_arena_member_.lock()->ScrollX());
  event_params_->SetValue("scrollY", gesture_arena_member_.lock()->ScrollY());
  event_params_->SetValue("isAtStart",
                          gesture_arena_member_.lock()->IsAtBorder(true));
  event_params_->SetValue("isAtEnd",
                          gesture_arena_member_.lock()->IsAtBorder(false));
  auto additional_params = GetEventParamsFromTouchEvent(touch_event);
  for (auto& param : *additional_params.Table()) {
    event_params_->SetValue(param.first, std::move(param.second));
  }
  return lepus::Value(std::move(event_params_));
}

void PanGestureHandler::OnBegin(float x, float y,
                                const std::shared_ptr<TouchEvent>& event) {
  if (!IsOnBeginEnable() || is_invoked_begin_) {
    return;
  }
  is_invoked_begin_ = true;
  SendGestureEvent(GestureConstants::ON_BEGIN, GetEventParamsInActive(event));
}

void PanGestureHandler::OnUpdate(
    float delta_x, float delta_y, const std::shared_ptr<TouchEvent>& event,
    const std::shared_ptr<GestureExtraBundle>& extra_bundle) {
  if (!IsOnUpdateEnable()) {
    return;
  }
  SendGestureEvent(GestureConstants::ON_UPDATE, GetEventParamsInActive(event));
}

void PanGestureHandler::OnStart(float x, float y,
                                const std::shared_ptr<TouchEvent>& event) {
  if (!IsOnStartEnable() || is_invoked_start_ || !is_invoked_begin_) {
    return;
  }
  is_invoked_start_ = true;
  SendGestureEvent(GestureConstants::ON_START, GetEventParamsInActive(event));
}

void PanGestureHandler::OnEnd(float x, float y,
                              const std::shared_ptr<TouchEvent>& event) {
  if (!IsOnEndEnable() || is_invoked_end_ || !is_invoked_begin_) {
    return;
  }
  is_invoked_end_ = true;
  SendGestureEvent(GestureConstants::ON_END, GetEventParamsInActive(event));
}

}  // namespace harmony
}  // namespace tasm
}  // namespace lynx
